package_scanner.nim

```diff
@@ -1,135 +1,367 @@
+  client = newHttpClient(timeout = httpTimeout) ## HTTP Client with Timeout
+  report = newJUnitOutputFormatter(openFileStream("report.xml", fmWrite)) ## JUnit Report XML
+addOutputFormatter(defaultConsoleFormatter())
+addOutputFormatter(report)
+when checkProfanity:
+  let profanities = filterIt(profanitiesStr, it.len > 0 and not(it.startsWith"#"))
+
+proc handler() {.noconv.} =
+  quit("CTRL+C Pressed, package_scanner is shutting down, Bye.")
+setControlCHook(handler)
+
+using url, name: string
 
-proc canFetchNimbleRepository(name: string, urlJson: JsonNode): bool =
-  # The fetch is a lie!
-  # TODO: Make this check the actual repo url and check if there is a
-  #       nimble file in it
-  result = true
-  var url: string
+func stripEndingSlash(url): string {.inline.} =
+  ## Strip the ending '/' if any else return the same string.
+  result = if url[^1] == '/': url[0 .. url.len - 2] else: url
 
-  if not urlJson.isNil:
-    url = urlJson.str
+func preprocessUrl(url, name): string =
+  ## Take **Normalized** URL & Name return Download link. GitHub & GitLab supported.
+  if likely(url.startswith("https://github.com/")):
+    result = url.replace("https://github.com/", "https://raw.githubusercontent.com/")
+    result = stripEndingSlash(result)
+    result &= "/master/" & name & ".nimble"
+  elif unlikely(url.startswith("https://0xacab.org/") or url.startswith("https://gitlab.")):
+    result = stripEndingSlash(result)
+    result &= "/raw/master/" & name & ".nimble"
 
+proc existsNimbleFile(url, name): string =
+  ## Take **Normalized** URL & Name try to Fetch the Nimble file. Needs SSL.
+  if url.startswith("http"):
     try:
-      discard getContent(url, timeout=10000)
-    except HttpRequestError, TimeoutError:
-      echo "W: ", name, ": unable to fetch repo ", url, " ",
-           getCurrentExceptionMsg()
-    except AssertionError:
-      echo "W: ", name, ": httpclient failed ", url, " ",
-           getCurrentExceptionMsg()
+      let urly = preprocessUrl(url, name)
+      doAssert urly.len > 0, "GIT or HG Hosting not supported: " & url
+      doAssert client.get(url).status == $Http200 # Check that Repo Exists.
+      result = urly
+    except TimeoutError, HttpRequestError, AssertionError:
+      warn("HttpClient request error fetching repo: " & url, getCurrentExceptionMsg())
     except:
-      echo "W: Another error attempting to request: ", url
-      echo "  Error was: ", getCurrentExceptionMsg()
+      warn("Unkown Error fetching repo: " & url, getCurrentExceptionMsg())
+  else:
+    result = url  # SSH or other non-HTTP kinda URLs?
 
-proc verifyAlias(pdata: JsonNode, result: var int) =
-  if not pdata.hasKey("name"):
-    echo "E: missing alias' package name"
-    result.inc()
+proc existsGitRepo(url): string =
+  ## Take **Normalized** URL try to Fetch the Git repo index page. Needs SSL.
+  if url.startswith("http"):
+    try:
+      doAssert client.get(url).status == $Http200 # Check that Repo Exists.
```

Comment:
Nitpick: in theory setting assertions and catching them is not supposed to be used for control flow and you can easily do if .. raise